### PR TITLE
Stagnation Point Pressure Constraint: physics-enforced Cp=1 at leading edge

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Add a soft physics constraint enforcing Cp approximately 1.0 at the leading-edge stagnation point. This extends the proven physics-constraint direction (Kutta TE: p_oodc -2.9% single seed) to the leading edge. At the stagnation point, the pressure coefficient Cp = 1 - (V/V_inf)^2 should be approximately 1.0 (velocity is near zero at the stagnation point). The model has no explicit signal forcing this — adding a soft constraint provides physics information that flows INTO the backbone via gradients, which our research shows is the winning approach (constraints > conditioning > replacement).

**For tandem configurations:** both fore-foil and aft-foil have stagnation points. The aft-foil stagnation point is affected by the fore-foil wake, making its location and pressure value harder to predict. Adding constraints at both stagnation points should help p_tan specifically.

**Literature:** This is a classical aerodynamics boundary condition. The stagnation point constraint is complementary to the Kutta TE constraint (PR #2374) — together they anchor the pressure distribution at both ends of each foil.

### Implementation

Add `--stagnation_constraint` and `--stagnation_weight` flags.

**Step 1: Identify stagnation points per foil.**

For each foil in a sample, find the stagnation node (the surface node where pressure should be maximum / velocity should be minimum). Two approaches (implement the simpler one):

**Approach A (geometry-based, preferred):**
- Use angle of attack to find the leading-edge node facing the freestream flow.
- Score each surface node: `score = x_node * cos(AoA) + y_node * sin(AoA)`. The node with the MINIMUM score is the leading-edge stagnation point (it faces into the flow).
- For tandem, identify fore vs aft foil nodes using the existing foil_id/foil_role field, and find one stagnation node per foil.

**Approach B (prediction-based, fallback):**
- Use the model's own pressure prediction to find the node with the highest predicted pressure per foil.
- Less robust (depends on model quality) but doesn't require AoA access.

**Step 2: Compute the stagnation loss.**

After the model predicts surface pressure:
```python
# p_pred: predicted pressure at surface nodes (in physical or asinh space)
# p_target: ground truth pressure at surface nodes
# stag_idx: index of stagnation node per foil

# Option 1: Direct pressure matching at stagnation point
# Use the ground truth pressure at the stagnation node as the target
L_stag = mean(|p_pred[stag_idx] - p_target[stag_idx]|)

# Option 2: Cp=1 constraint (if you can convert to Cp)
# Cp = (p - p_inf) / (0.5 * rho * Umag^2)
# L_stag = mean(|Cp_pred[stag_idx] - 1.0|)
```

Use Option 1 (direct pressure matching at stag node) since it avoids needing to convert between pressure spaces. The stagnation constraint becomes: "predict the stagnation pressure accurately."

**Step 3: Weight the aft-foil stagnation higher.**

For tandem samples, the aft-foil stagnation is harder (affected by wake). Apply 2x weight:
```python
L_stag = w_fore * |p_pred[stag_fore] - p_target[stag_fore]| + 
         w_aft * |p_pred[stag_aft] - p_target[stag_aft]|
# where w_aft = 2 * w_fore
```

**Step 4: Add to total loss.**
```python
total_loss = main_loss + stagnation_weight * L_stag
```

### Training Setup

```bash
cd cfd_tandemfoil && python train.py \
  --agent nezuko --wandb_name "nezuko/stag-constraint-w05-s42" \
  --wandb_group stagnation-constraint --seed 42 \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --stagnation_constraint --stagnation_weight 0.5
```

Run 2 seeds (42, 73) with weight=0.5. If results show promise but the weight is too aggressive (regressions on other metrics), try weight=0.1 as well.

### What to Watch

- **p_tan** primary target — aft-foil stagnation constraint directly addresses tandem prediction difficulty
- **p_in** and **p_re** should benefit from better leading-edge pressure anchoring
- If the stagnation node identification is noisy, the constraint becomes harmful — monitor whether the model converges

## Baseline

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| **p_in** | **11.872** | < 11.872 |
| p_oodc | 7.459 | < 7.459 |
| **p_tan** | **26.319** | < 26.319 |
| **p_re** | **6.229** | < 6.229 |

Baseline PR: #2357. W&B: aycq1m8m (s42), 9sk276v6 (s73).

Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Metrics vs Baseline (weight=0.5, 2 seeds)

| Metric | Baseline | Seed 42 | Seed 73 | **2-seed avg** | Δ |
|--------|----------|---------|---------|----------------|---|
| **p_in** | 11.872 | 13.128 | 12.730 | **12.929** | +1.057 (+8.9%) ❌ |
| p_oodc | 7.459 | 8.003 | 8.091 | **8.047** | +0.588 (+7.9%) ❌ |
| **p_tan** | 26.319 | 28.659 | 28.446 | **28.553** | +2.234 (+8.5%) ❌ |
| **p_re** | 6.229 | 6.763 | 7.049 | **6.906** | +0.677 (+10.9%) ❌ |

All metrics degraded uniformly by ~8-11%. Stagnation constraint with weight=0.5 did not help.

### Run details
- Seed 42 W&B: `a3s57yqu` (nezuko/stag-constraint-w05-s42)
- Seed 73 W&B: `b63w3haj` (nezuko/stag-constraint-w05-s73)
- Group: `stagnation-constraint`
- ~145-150 training epochs (3h timeout)

### Commands used
```
# Seed 42
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py --agent nezuko --wandb_name "nezuko/stag-constraint-w05-s42" --wandb_group stagnation-constraint --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 --stagnation_constraint --stagnation_weight 0.5

# Seed 73: same with --seed 73 on CUDA_VISIBLE_DEVICES=1
```

## What happened

**Negative result.** All metrics degraded uniformly (~8-11%), including p_in (single-foil) and p_re — which should be completely unaffected by the aft-foil 2x weighting. This suggests the stagnation constraint is adding noise/gradient conflict across all samples, not just improving aft-foil predictions.

**Why it likely failed:**

1. **Stagnation node identification is fragile**: The geometry-based approach () finds the node facing most into the freestream. However, for highly curved or irregular meshes, this geometric minimum may not correspond to the true fluid-dynamic stagnation point. The true stagnation point shifts with boundary layer effects and is not exactly at the geometric leading edge.

2. **Weight 0.5 is too aggressive**: The stagnation loss at weight=0.5 competes strongly with the main physics loss. The model is being pulled toward matching pressure at a single point per foil, at the cost of the overall pressure distribution. A much smaller weight (0.05-0.1) might not cause regression.

3. **Uniform degradation across all splits**: The fact that p_in and p_re regressed (not just p_tan) indicates the constraint is harming single-foil convergence too — even though single-foil samples only have the fore-foil stagnation constraint. This suggests the stagnation node identification is inconsistent or the pressure at the identified nodes is noisy.

4. **y_norm-space matching**: The loss is computed in normalized prediction space ( vs ). After asinh_pressure transformation and per-sample std normalization, the stagnation node pressure may not have a stable target value — it varies widely depending on sample normalization, making the constraint signal noisy.

5. **PCGrad interaction**: The stagnation loss is added to all PCGrad task groups as a shared term. If the stagnation loss gradient conflicts with the main task gradient directions, PCGrad projection could be suppressing useful gradients.

## Note on weight=0.1
The PR mentions trying weight=0.1 if weight=0.5 is too aggressive. Given that:
- Uniform degradation across ALL splits (not just tandem) suggests the stagnation node identification itself is problematic
- p_in regressed by +8.9%, which has no tandem interaction at all

The issue is likely identification quality + normalization-space mismatch rather than just weight magnitude. I did NOT run weight=0.1 separately as the evidence points to a fundamental issue with the approach in this implementation. The advisor may want to instruct a weight=0.1 run anyway.

## Suggested follow-ups

1. **Use prediction-based stagnation node identification** (Approach B from the PR): Find the node with the highest predicted pressure per foil each batch, rather than geometry-based. This adapts to the true fluid stagnation location and avoids geometry approximation errors.

2. **Reduce weight significantly**: Try weight=0.05 or 0.1. The Kutta TE constraint (PR #2374) succeeded at very small weights — similar caution may be needed here.

3. **Apply constraint in original (physical) pressure space** rather than normalized prediction space: Convert pred and target back to physical pressure before computing the constraint loss. The normalized space has per-sample std scaling that makes stagnation pressure values unstable as constraint targets.

4. **Only apply to tandem samples**: If the goal is specifically p_tan improvement, restrict the constraint to tandem samples only — this avoids degrading single-foil performance.

5. **Soft upper bound rather than exact match**: Instead of , use  — only penalize if the model UNDER-predicts stagnation pressure. Stagnation pressure being the maximum naturally has a one-sided constraint.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2386 because ADVISOR: Closing — uniform 8-11% regression across ALL splits. Good analysis, Nezuko. The uniform degradation (including p_in, which has no tandem interaction) confirms this isn't just weight tuning — the stagnation node identification is fragile (geometric LE ≠ fluid stagnation point) and the constraint target in normalized space is noisy. Combined with PCGrad gradient conflicts, the approach harms convergence globally. The physics constraint lesson continues: wall constraints must be applied carefully in the right space and at the right location. Nezuko is already on #2396 (ANP Context Dropout) — full speed ahead on the p_re fix campaign..
